librepo/downloader.c

```diff
@@ -691,64 +691,77 @@ select_suitable_mirror(LrDownload *dd,
     assert(!err || *err == NULL);
 
     *selected_mirror = NULL;
+    // mirrors_iterated is used to allow to use mirrors multiple times for a targed
+    int mirrors_iterated = 0;
+    //  Iterate over mirrors for a target. When all mirrors cannot be used for a download at the
+    //  time during the first iteration, allow additional iterataions to get number af allowed
+    //  failures equal to dd->allowed_mirror_failures
+    do {
+        // Iterate over mirror for the target
+        for (GSList *elem = target->lrmirrors; elem; elem = g_slist_next(elem)) {
+            LrMirror *c_mirror = elem->data;
+            gchar *mirrorurl = c_mirror->mirror->url; // shortcut
+
+            if (mirrors_iterated == 0) {
+                if (g_slist_find(target->tried_mirrors, c_mirror)) {
+                    // This mirror was already tried for this target
+                    continue;
+                }
+                if (c_mirror->successful_transfers == 0 &&
+                    dd->allowed_mirror_failures > 0 &&
+                    c_mirror->failed_transfers >= dd->allowed_mirror_failures)
+                {
+                    // Skip bad mirrors
+                    g_debug("%s: Skipping bad mirror (%d failures and no success): %s",
+                            __func__, c_mirror->failed_transfers, mirrorurl);
+                    continue;
+                }
+            } else if (mirrors_iterated < c_mirror->failed_transfers) {
```

Comment:
So IIUC, mirrors can pick up `failed_transfers` on different targets that were downloaded previously and can thus always end up being skipped for the subsequent downloads.

Example:
I have one mirror and two targets. The first target is not on the mirror (404), the second is there. The first target will be attempted 4 times and raise `c_mirror->failed_transfers` to 4. For the second target the mirror will never be tried, because the conditions in the code above will always rule it out. Is this correct?

Reply:
Yes, it is correct.

Reply:
Well, then it seems like a problem with the algorithm, as it does not try every mirror at least once for each target, which I think it should?

Reply:
The algorithm is correct, because there are two contradicting requirements:
  a. try to download everything
  b. fail fast